collections/collections.go

```diff
@@ -0,0 +1,83 @@
+package collections
+
+import (
+	"errors"
+	"math"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+var (
+	// ErrNotFound is returned when the provided key is not present in the StorageProvider.
+	ErrNotFound = errors.New("collections: not found")
+	// ErrEncoding is returned when something fails during key or value encoding/decoding.
+	ErrEncoding = errors.New("collections: encoding error")
+)
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// Prefix defines a segregation namespace
+// for specific collections objects.
+type Prefix struct {
+	raw []byte // TODO(testinginprod): maybe add a humanized prefix?
+}
+
+// Bytes returns the raw Prefix bytes.
+func (n Prefix) Bytes() []byte { return n.raw }
+
+// NewPrefix returns a Prefix given the provided namespace identifier.
+// Prefixes of the same module must not start
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
+		}
+		prefix = []byte{uint8(c)}
+	case string:
+		prefix = []byte(c)
+	case []byte:
+		prefix = c // maybe copy?
+	}
+	return Prefix{raw: prefix}
+}
+
+// KeyEncoder defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collections keys.
+type KeyEncoder[T any] interface {
```

Comment:
What do you think about calling these `Codec`s?

Reply:
Sounds good to me.